### PR TITLE
Add topologies test to integration test suite in TP

### PR DESCRIPTION
**Pull request by dpham692**
## What does this PR (Pull Request) do?

This PR created to add Topologies test base to the integration tests suite. More test will be added to this base.

## Which Traffic Control components are affected by this PR?

- CDN in a Box
- Traffic Portal
- CI tests

## What is the best way to verify this PR?

In traffic_portal/integration directory,  run the traffic portal tests and make sure all the test pass.
Change the spec name to **"specs/Topologies.spec.js"** in **config.ts** file to run the specific tests.
## The following criteria are ALL met by this PR

- [ ] This PR includes tests OR I have explained why tests are unnecessary
- [ ] This PR includes documentation OR I have explained why documentation is unnecessary
- [ ] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [ ] This PR includes any and all required license headers
- [ ] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: traffic_portal/test/integration/specs/Topologies.spec.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    })
+})
+
+describe('Clean up prereq and test data for Topologies Test', function () {
+    it('Cleanup', async function () {
+        let cleanupData = JSON.parse(readFileSync(cleanupFile,'utf-8'));
+        let output = await api.UseAPI(cleanupData);
+        expect(output).toBeNull();
+    })
+})
```
Comment (ocket8888) [inline traffic_portal/test/integration/specs/Topologies.spec.ts:77]: missing newline at EOF

File: traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        result = await basePage.GetOutputMessage().then(function (value) {
+            if (topologies.validationMessage == value) {
+                return true;
+            } else {
+                return false;
+            }
+        })
+        return result;
+    }
+}
```
Comment (ocket8888) [inline traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts:192]: Missing newline at EOF

File: traffic_portal/test/integration/specs/Topologies.spec.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { browser } from 'protractor';
+import { LoginPage } from '../PageObjects/LoginPage.po';
+import { TopologiesPage } from '../PageObjects/TopologiesPage.po';
+import { TopNavigationPage } from '../PageObjects/TopNavigationPage.po';
+import  * as using  from "jasmine-data-provider";
```
Comment (ocket8888) [inline traffic_portal/test/integration/specs/Topologies.spec.ts:24]: `import * as using from "jasmine-data-provider"` won't work, I think. I believe it has to be `import using from "jasmine-data-provider"`.

**Comment by dpham692** [comment]
@ocket8888  I fixed all the comment and change the tests to new format.

File: traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        await snp.NavigateToTopologiesPage();
+        await this.txtSearch.clear();
+        await this.txtSearch.sendKeys(name);
+        if (await browser.isElementPresent(element(by.xpath("//td[@data-search='^" + name + "$']"))) == true) {
+            await element(by.xpath("//td[@data-search='^" + name + "$']")).click();
+            return true;
+        } 
+        return false;
+    }
+    async DeleteTopologies(topologies: Topologies):Promise<Boolean>{
```
Comment (ocket8888) [inline traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts:89]: `Boolean` is the type of the constructor-used-as-a-function `globalThis.Boolean` - this will compile just fine because `boolean` is assignable to `Boolean` and the caller can check the truthiness of a `Boolean`. However, a `Boolean` is an `object`, but a `boolean` isn't. There are lots of subtle differences because of that, but mainly if you later use typeguards to check something's type like `if (typeof x === "boolean")`, you're gonna be fooled, because `typeof(false)` outputs `"boolean"` while `typeof(new Boolean())` actually outputs `"object"`.

This playground helps illustrate: https://www.typescriptlang.org/play?strict=true#code/DYUwLgBAQgXND29QEMB2EC8EwCcCuIA3ALABQokARnJYiulrgSaWQGZ6oDGYAlvOgAOOXqjABJMAAp4ceJQBWIHgEo4AN3i8AJhADeZCEYhcBAZyQgAdMHgBzGSpYBfMmWGiJ0qE-cixklKUvqykUJgQqCAA7giWaFIhlBFRsVB0IAlOQA

tl;dr use `boolean` not `Boolean` (and likewise `number` not `Number`, `string` not `String`, `object` not `Object`, `function` (or better yet strict call signature definitions) not `Function` etc.)

**Review by ocket8888 — COMMENTED** [review]
Why was the `package-lock.json` file changed? It doesn't look like you actually added or removed any dependencies, since `package.json` was unchanged

File: traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts (hunk 4 above)
Comment (ocket8888) [reply traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts:89]: For further confusion, `Boolean` when used as a function spits out `boolean`s, not `Boolean`s. For example:
```javascript
typeof(Boolean()); // "boolean"
typeof(new Boolean()); // "object"
Boolean() instanceof Boolean; // false
Boolean instanceof Boolean; // false
new Boolean() instanceof Boolean; // true;
```

Comment (dpham692) [reply traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts]: Nice catch! The package lock change I think because when I merged or npm install.

**Comment by dpham692** [comment]
Finish fixing test for Docker environment

File: traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        } 
+        return false;
+    }
+    async DeleteTopologies(topologies: DeleteTopologies):Promise<boolean>{
+        let name = topologies.Name + this.randomize;
+        let basePage = new BasePage();
+        await this.btnDelete.click();
+        await this.txtConfirmName.sendKeys(name);
+        await basePage.ClickDeletePermanently();
+        return await basePage.GetOutputMessage().then(value => value === topologies.validationMessage);
```
Comment (ocket8888) [inline traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts:99]: This class extends `BasePage`, so couldn't you do `this.ClickDeletePermanently();` etc. instead of making a new `BasePage` instance?

Comment (dpham692) [reply traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts]: Okie. Fixed.

File: traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    async OpenTopologiesPage(){
+        let snp = new SideNavigationPage();
+        await snp.NavigateToTopologiesPage();
+    }
+    async OpenTopologyMenu(){
+        let snp = new SideNavigationPage();
+        await snp.ClickTopologyMenu();
+    }
+    public async CreateTopologies(topologies: Topologies): Promise<boolean> {
+        let basePage = new BasePage();
```
Comment (ocket8888) [inline traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts:56]: I don't think you need this BasePage instance; you're inheriting BasePage already

File: traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        await this.txtSearch.sendKeys(name);
+        if (await browser.isElementPresent(element(by.xpath("//td[@data-search='^" + name + "$']"))) == true) {
+            await element(by.xpath("//td[@data-search='^" + name + "$']")).click();
+            return true;
+        } 
+        return false;
+    }
+    async DeleteTopologies(topologies: DeleteTopologies):Promise<boolean>{
+        let name = topologies.Name + this.randomize;
+        let basePage = new BasePage();
```
Comment (ocket8888) [inline traffic_portal/test/integration/PageObjects/TopologiesPage.po.ts:95]: same as above RE: `BasePage`

**Review by ocket8888 — COMMENTED** [review]
I've heard other people had problems trying to use inherited methods instead of new `BasePae` instances; if you can't get it to work then just don't worry about it. It's something I can look into in my own time if it bothers me that much and isn't trivial to do.

**Comment by dpham692** [comment]
Yeah, look like when I revert back to basePage it work. I will figure out why the basePage fail and do on other PR.

**Review by ocket8888 — APPROVED** [review]
